examples/golang/Dockerfile-proxy

```diff
@@ -0,0 +1,5 @@
+FROM mosnio/envoy-go-extension:beta-v1
```

Comment:
Why is this pulling from mosnio? this should pull from envoy.

Reply:
> Why is this pulling from mosnio? this should pull from envoy.

Because we put the implementation of the go extension API in this image, so that users olny need to provide the go code to work, just like this [example](https://github.com/envoyproxy/envoy/tree/57ffc1cab6fb151f47f064e829bd924b0b9737a9/examples/golang/envoy-go-filter-samples/pkg/plugins/simple).  We could be pushed implementation after the community agrees on the design and completes the API. Just like the contribution of [MetaProtocol Proxy API](https://github.com/envoyproxy/envoy/pull/18823).

Reply:
Ultimately the code needs to live in this repo. So if you want to iterate on it more let's just hold this PR until you get some more consensus.

/wait

Reply:
@mattklein123   I've already committed the API and implementation to this PR. The sandbox example will be pushed after the PR merge because it depends on the image `envoyproxy/envoy-contrib-dev:latest` which contains the go extension feature.

Reply:
i think its better to push the sandbox separately (it requires documentation and verification etc) but i should say that the `envoy-contrib-dev` in ci should already have the go extension

Reply:
> i think its better to push the sandbox separately (it requires documentation and verification etc) but i should say that the `envoy-contrib-dev` in ci should already have the go extension

@phlax  Yes, the sandbox will be pushed in another PR.